stan/math/prim/prob/categorical_logit_lpmf.hpp

```diff
@@ -7,62 +7,63 @@
 #include <stan/math/prim/fun/log_softmax.hpp>
 #include <stan/math/prim/fun/log_sum_exp.hpp>
 #include <stan/math/prim/fun/sum.hpp>
+#include <stan/math/prim/fun/size_mvt.hpp>
+#include <stan/math/prim/fun/to_ref.hpp>
+#include <stan/math/prim/fun/vector_seq_view.hpp>
+#include <stan/math/prim/fun/scalar_seq_view.hpp>
 #include <vector>
 
 namespace stan {
 namespace math {
 
-// CategoricalLog(n|theta)  [0 < n <= N, theta unconstrained], no checking
-template <bool propto, typename T_prob,
-          require_eigen_col_vector_t<T_prob>* = nullptr>
-return_type_t<T_prob> categorical_logit_lpmf(int n, const T_prob& beta) {
+/** \ingroup prob_dists
+ * Returns the log PMF of the categorical distribution with a softmax
+ * inverse link function. If containers of integers and/or probabilities
+ * are supplied, returns the log sum of the PMF.
+ *
+ * @tparam T_n type of integer parameters
+ * @tparam T_prob type probability vector(s)
+ * @param n integer parameter(s)
+ * @param beta probability vector(s)
+ * @return log probability
+ */
+template <bool propto, typename T_n, typename T_prob>
+return_type_t<T_prob> categorical_logit_lpmf(const T_n& n, const T_prob& beta) {
   static const char* function = "categorical_logit_lpmf";
-  check_bounded(function, "categorical outcome out of support", n, 1,
-                beta.size());
+
   ref_type_t<T_prob> beta_ref = beta;
-  check_finite(function, "log odds parameter", beta_ref);
+  scalar_seq_view<T_n> n_vec(n);
+  vector_seq_view<ref_type_t<T_prob>> beta_vec(beta_ref);
+
+  size_t vec_size = std::max(size(n), size_mvt(beta_ref));
 
-  if (!include_summand<propto, T_prob>::value) {
+  if (!include_summand<propto, T_prob>::value || size(n) == 0) {
     return 0.0;
   }
 
-  // FIXME:  wasteful vs. creating term (n-1) if not vectorized
-  return beta_ref.coeff(n - 1)
-         - log_sum_exp(beta_ref);  // == log_softmax(beta)(n-1);
-}
-
-template <bool propto, typename T_prob,
-          require_eigen_col_vector_t<T_prob>* = nullptr>
-return_type_t<T_prob> categorical_logit_lpmf(const std::vector<int>& ns,
-                                             const T_prob& beta) {
-  static const char* function = "categorical_logit_lpmf";
-
-  check_bounded(function, "categorical outcome out of support", ns, 1,
-                beta.size());
-  ref_type_t<T_prob> beta_ref = beta;
-  check_finite(function, "log odds parameter", beta_ref);
-
-  if (!include_summand<propto, T_prob>::value) {
-    return 0.0;
+  if (size_mvt(beta_ref) > 1) {
+    check_consistent_sizes(function, "Integers", n, "Probabilities", beta_ref);
```

Comment:
Alright I think this is the right check.

So `check_consistent_sizes` is the vectorized lpdf check. It says arguments either be size 1, or all have the same size.

The problem here is if `beta_ref` is an `Eigen::VectorXd`, it should count as size 1 (that's the `size_mvt` logic). So because `check_consistent_sizes` does size instead of `size_mvt`, it doesn't work out of the box.

However, the logic does work if we check that `size_mvt(beta) > 1`. In this case `size(beta)` and `size_mvt(beta)` are the same thing.

If `size_mvt(beta) == 1)`, then we're good.

Something that bothered me about these checks is the error message is not great:

```
categorical_lpmf: Integers has size = 3, but Probabilities has size 4; and they must be the same size.
```

Changing `Probabilities` to `probabilities parameter` would help. The `Integers` thing should probably change too.